Add example11: Python JobSpec Submit

cmoussa1 opened the pull request:
This PR looks to add another example to our `flux-workflow-examples` repository based on the discussion had in issue [#573](https://github.com/flux-framework/flux-sched/issues/573) in `flux-framework/flux-sched`. This example uses @grondo's **bulksubmit.py** script, which submits a large directory of jobpsec files for `sleep 0` jobs, then keeps track of its progress with a visual progress bar:

```
./bulksubmit.py jobs
bulksubmit: Starting...
bulksubmit: submitted 1025 jobs in 2.87s. 357.20job/s
bulksubmit: First job finished in about 2.897s
|██████████████████████████████████████████████████████████| 100.0% (29.1 job/s)
bulksubmit: Ran 1025 jobs in 35.2s. 29.1 job/s
```

grondo commented:
Thanks! I tweaked the `bulksubmit.py` script slightly to be slightly more usable. This version might be better for this example

```python
import time
import sys
import flux
from flux import job
from flux import constants

t0 = time.time()

flags = flux.constants.FLUX_JOB_WAITABLE
h = flux.Flux()
jobs = []

label="bulksubmit"

def log(s):
    print(label + ": " + s)

def progress(fraction, length=72, suffix=""):
    fill = int(round(length * fraction))
    bar = '\u2588' * fill + '-' * (length - fill)
    s = '\r|{0}| {1:.1f}% {2}'.format(bar, 100*fraction, suffix)
    sys.stdout.write(s)
    if fraction == 1.:
        sys.stdout.write('\n')

def submit_cb(f, arg):
    jobs.append (job.submit_get_id (f))

log("Starting...")
for file in sys.argv[1:]:
    with open(file) as jobspec:
        job.submit_async(h, jobspec.read(), flags=flags).then(submit_cb)

if h.reactor_run(h.get_reactor(), 0) < 0:
    h.fatal_error("reactor start failed")

# print(jobs)

total = len(jobs)
dt = time.time() - t0
jps = len(jobs)/dt
log("submitted {0} jobs in {1:.2f}s. {2:.2f}job/s".format(total, dt, jps))

count = 0
while count < total:
    job.wait(h)
    count = count + 1
    if count == 1:
        log("First job finished in about {0:.3f}s".format(time.time() - t0))
    suffix = "({0:.1f} job/s)".format(count/(time.time() - t0))
    progress(count/total, length=58, suffix=suffix)

dt = time.time() - t0
log("Ran {0} jobs in {1:.1f}s. {2:.1f} job/s".format(total, dt, total/dt))
```

Main difference is it takes a list of jobspec on the commandline instead of a directory:
```console
$ src/cmd/flux start -s 2 flux python bulksubmit.py  jobs/*.json
bulksubmit: Starting...
bulksubmit: submitted 1024 jobs in 1.49s. 685.03job/s
bulksubmit: First job finished in about 2.101s
|██████████████████████████████████████████████████████████| 100.0% (26.1 job/s)
bulksubmit: Ran 1024 jobs in 39.2s. 26.1 job/s
```

grondo commented:
Oh, also, I don't think this script works with python2 so I wonder if `#!/usr/bin/env python3` (or something) should be added to the file.

cmoussa1 commented:
> Oh, also, I don't think this script works with python2 so I wonder if `#!/usr/bin/env python3` (or something) should be added to the file.

Yes! Good catch. While I was trying to run your script when creating the example I was getting errors because I was using python2. I included `#!/usr/bin/env python3` in the **bulksubmit.py** file, but maybe an additional note about using python3 in the README would be helpful too?

grondo commented:
Oh yeah, sorry I just didn't notice you had already made the script python3 specific. Sorry!

cmoussa1 commented:
Thanks for upgrading your **bulksubmit.py** script @grondo! I went ahead and added those updates and force-pushed the fixes to both the Python script and **README.md**.

grondo commented on example11/README.md:
```diff
@@ -0,0 +1,50 @@
+
+``for i in `seq 1 1024`; do cp jobs/0.json jobs/${i}.json; done``
+
+6. Run the **bulksubmit.py** script and pass the **jobs** directory as an argument:
```
and pass all jobspec in the `jobs` directory as an argument with a shell glob `jobs/*.json`

grondo commented on example11/README.md:
```diff
@@ -0,0 +1,50 @@
+
+- `h = flux.Flux()` creates a new Flux handle which can be used to connect to and interact with a Flux instance.
+
+- `job_submit_async(h, jobspec.read(), flags=flags)` submits a jobspec to be run.
```
Also should describe that `job_submit_async` submits a jobspec, returning a future which will be fulfilled when the submission of this job is complete.

`.then(submit_cb)`, called on the returned future, will cause our callback `submit_cb()` to be invoked when the submission of this job is complete and a jobid is available.  To process job submission RPC responses and invoke callabacks, the flux reactor for handle `h` must be run:

```python
if h.reactor_run(h.get_reactor(), 0) < 0:
￼    h.fatal_error("reactor start failed")
```

The reactor will return automatically when there are no more outstanding RPC responses, i.e. all jobs have been submitted.

grondo commented on example11/README.md:
```diff
@@ -0,0 +1,50 @@
+
+- `job.wait(h)` waits for a job to transition to the **INACTIVE** state.
+
+- The following creates a reactor to monitor for events on the broker handle:
```
Minor: It might be more clear to say `to process events` instead of `monitor for events` since the latter might be confused with event messages and not event loop events. However, if you move the description of the reactor as suggested above, this might just go away.

grondo commented on example11/README.md:
```diff
@@ -0,0 +1,50 @@
+
+- `job_submit_async(h, jobspec.read(), flags=flags)` submits a jobspec to be run.
+
+- `job.wait(h)` waits for a job to transition to the **INACTIVE** state.
```
Slight correction for clarity:

`job.wait(h)` will wait for *any* job submitted with the `FLUX_JOB_WAITABLE` flag to transition to the **INACTIVE** state.

cmoussa1 commented:
Thanks for your feedback and suggestions @grondo - it really helps make the example look a lot more comprehensive! I went ahead and pushed those fixes up.

cmoussa1 commented:
Rebased to catch up after PR #61

SteVwonder commented:
PS - Can you also add this example to the top-level README in the same PR?  It can be a separate commit if you want.

cmoussa1 commented:
Thanks for reminding me @SteVwonder! I went ahead and rebased and pushed a commit that adds this example to our top level **README**. I renamed the newest example to `async-bulk-job-submit`, which is ordered at number 4, after `job-submit-wait`.

SteVwonder commented:
> Edit: Oh, also, I just realized this branch is still named example11. If needed I can rename it to something clearer since we are no longer naming our examples by numbers...

The branch name doesn't end up in the main fork history in any way as far as I understand.  So no need.  I think it is fine as is.

Now that #64 has been merged, can you do a quick rebase of this PR onto master?  Then I'll push the button.

Thanks for all the hard work on these @cmoussa1!

cmoussa1 commented:
Sure thing! Just rebased to catch up after PR #64

cmoussa1 commented:
thanks! 🎉
